### PR TITLE
Bugfix/convolve 2d overlap 115

### PR DESCRIPTION
**Pull request by richpsharp**
Fixing an issue introduced by optimization steps in 2.1.0. Also added a test that would have caught this bug in the 2.1.0 (tested on 2.1.0 and it fails, tested on this branch and it passes).

### CODE REVIEW

File: tests/test_geoprocessing.py [hunk 1]
```diff
@@ -4186,3 +4186,40 @@ def test_convolve_2d_nodata(self):
...
+        """PGP.geoprocessing: test convolve 2d with large gaussian kernel."""
+        # choosing twice the max memory block
+        n_pixels = 256*2
+        # this is a fun seed random seed
+        numpy.random.seed(n_pixels)
+        signal_array = numpy.random.random((n_pixels, n_pixels))
+
+        workspace_dir = 'local'
+        if not os.path.exists(workspace_dir):
+            os.makedirs(workspace_dir)
```
Comment (dcdenu4) [inline tests/test_geoprocessing.py]: Should this be `self.workspace_dir`? so that it gets deleted on cleanup?

File: tests/test_geoprocessing.py [hunk 2]
```diff
@@ -4186,3 +4186,40 @@ def test_convolve_2d_nodata(self):
...
+            (signal_path, 1), (kernel_path, 1), target_path,
+            ignore_nodata_and_edges=False, mask_nodata=True)
+        target_array = pygeoprocessing.raster_to_numpy_array(target_path)
+
+        # gaussian filter with constant is the same as bleeding off the edges
+        expected_output = scipy.ndimage.gaussian_filter(
+            signal_array, 1.0, mode='constant')
+
+        _array_to_raster(
+            expected_output, None, os.path.join(workspace_dir, 'extected.tif'))
```
Comment (dcdenu4) [inline tests/test_geoprocessing.py]: Not a big deal, but was this supposed to be `expected.tif`?

Comment (dcdenu4) [reply tests/test_geoprocessing.py]: Also, is there any reason to save this array to a raster?

File: tests/test_geoprocessing.py [hunk 3]
```diff
@@ -4186,3 +4186,40 @@ def test_convolve_2d_nodata(self):
...
+
+        target_path = os.path.join(workspace_dir, 'target.tif')
+        pygeoprocessing.convolve_2d(
+            (signal_path, 1), (kernel_path, 1), target_path,
+            ignore_nodata_and_edges=False, mask_nodata=True)
+        target_array = pygeoprocessing.raster_to_numpy_array(target_path)
+
+        # gaussian filter with constant is the same as bleeding off the edges
+        expected_output = scipy.ndimage.gaussian_filter(
+            signal_array, 1.0, mode='constant')
```
Comment (dcdenu4) [inline tests/test_geoprocessing.py:4219]: This is pretty slick, basically created a convolution test case that mimics a scipy guassian filter process and test against it.

File: tests/test_geoprocessing.py [hunk 4]
```diff
@@ -4186,3 +4186,40 @@ def test_convolve_2d_nodata(self):
...
         expected_output[-1, -1] = 1
         numpy.testing.assert_allclose(target_array, expected_output)
+
+    def test_convolve_2d_gaussian(self):
+        """PGP.geoprocessing: test convolve 2d with large gaussian kernel."""
+        # choosing twice the max memory block
+        n_pixels = 256*2
+        # this is a fun seed random seed
+        numpy.random.seed(n_pixels)
+        signal_array = numpy.random.random((n_pixels, n_pixels))
```
Comment (dcdenu4) [inline tests/test_geoprocessing.py]: Nice. It does appear that Numpy frowns upon reseeding the BitGenerator (https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html) but, it's probably fine. Also, I'm not sure what numpy version we're sticking to at the moment. It looks like they've recently reworked their random API.

File: src/pygeoprocessing/geoprocessing.py [hunk 5]
```diff
@@ -2540,7 +2541,6 @@ def _fill_work_queue():
     # indexes
     LOGGER.debug('start worker thread')
     write_queue = queue.Queue(10)
-    worker_list = []
```
Comment (dcdenu4) [inline src/pygeoprocessing/geoprocessing.py:2543]: This was left over from the last changes and not removed I take it?

File: src/pygeoprocessing/geoprocessing.py [hunk 6]
```diff
@@ -2554,7 +2554,6 @@ def _fill_work_queue():
 
     # this set is indexed by (x,y) tuple from iterblock offsets to check if the
     # target block is on its first pass and set to zero if so.
```
Comment (dcdenu4) [inline src/pygeoprocessing/geoprocessing.py]: Can this comment go as well?

**Review by dcdenu4 — CHANGES_REQUESTED** [review]
Hey @richpsharp thanks for this fix. I tested it on my data and it works. I just have some minor questions, suggestions, and comments!

File: tests/test_geoprocessing.py (hunk 1 above)
Comment (richpsharp) [reply tests/test_geoprocessing.py]: Oh gosh of course it should, I do this a lot when I'm debugging tests and then I also forget to change it back. Changed in 3349d4b9eb4ba5d09fc8ba576e8fb13e00c8e657

File: tests/test_geoprocessing.py (hunk 2 above)
Comment (richpsharp) [reply tests/test_geoprocessing.py]: Nope! Same as above, left over from testing/debugging. Removed in 3a5f174093f133daeb8f60de269d8220a2341971

File: tests/test_geoprocessing.py (hunk 3 above)
Comment (richpsharp) [reply tests/test_geoprocessing.py:4219]: Thanks! Trying to get some confidence!

File: tests/test_geoprocessing.py (hunk 4 above)
Comment (richpsharp) [reply tests/test_geoprocessing.py]: I did not know about this! I guess I get why they want to do that -- no global state? I changed it in ad2bf39ea328fe9d71249e1f86f8aefd4ff7a0ed

File: src/pygeoprocessing/geoprocessing.py (hunk 5 above)
Comment (richpsharp) [reply src/pygeoprocessing/geoprocessing.py:2543]: Yes!

File: src/pygeoprocessing/geoprocessing.py (hunk 6 above)
Comment (richpsharp) [reply src/pygeoprocessing/geoprocessing.py]: Yes, removed in 8afbb2e387c996f4905c3edf02a6802596e2c334

**Comment by richpsharp** [comment]
> Hey @richpsharp thanks for this fix. I tested it on my data and it works. I just have some minor questions, suggestions, and comments!

Thanks @dcdenu4 Dough, those were good comments. I addressed all of them and pushed and re-requesting your review.

**Review by dcdenu4 — APPROVED** [review]
Thanks Rich! Looks good to me.

**Pull request merged**